README
This is Boa, a high performance web server for Unix-alike computers,
covered by the Gnu General Public License.  This is version 0.92r,
released 11 February, 2000.  It is well tested and appears to be of
production quality, given some limitations spelled out in the
documentation.  There is a new production branch of Boa, the 0.94
series, that should be used instead of this one for most purposes.
See the www.boa.org web site for more information.

Boa was created by Paul Phillips <psp@well.com>.  It is now being
maintained and enhanced by Larry Doolittle <ldoolitt@jlab.org>
and Jon Nelson <nels0988@tc.umn.edu>.

For more information (including installation instructions) examine
the files docs/*.txt, point your web browser to docs/readme.html, or
visit the Boa homepage at

        http://www.boa.org/



src/Changelog
** Changes from v0.92q to v0.92r **
 * Fix potential buffer overruns in alias.c by changing sprintf to snprintf
   (e-mail support@boa.org if you can't find snprintf for your computer)
 * Add O_EXCL to temp file open in read.c
 * Check for out-of-memory strdup in read.c
 * Replaced .gif files with .png
 * Updated e-mail addresses, referenced boa-0.94 in README

** Changes from v0.92p to v0.92q **
 * fix 1 byte overrun on path_translated in get.c
 * more robust input handling for if-modified-since
 * bring all preprocessor "#"'s to column 1

** Changes from v0.92o to v0.92p **

 * Documented misbehavior of CGI, SIGHUP, short aliases, stale dircache.
 * Documented how to patch signals.c for use on SunOS.
 * Closed file descriptor leak when redirecting a bare directory URL to
 one with an appended "/".
 * Closed potential file descriptor leak if errors encountered generating
 on-the-fly index.
 * Cleaned up include file handling to be simultaneously compatible with
 Linux, SunOS, HP-UX, and AIX.
 * Supress message body for codes 302, 400, 403, 404, 500, and 501 if
 incoming request is "HEAD".

** Changes from v0.91 to v0.92o **
 (0.92o released 27 December, 1996)

 * Maintenance handover from Paul Phillips to Larry Doolittle
 * Changed (char)NULL to '\0'
 * Cleaned up signal handler prototypes in signals.c
 * Modified handling of CGI environment variable PATH_TRANSLATED,
 should now work the same as NCSA.
 * More conservative buffer size in add_cgi_env()
 * Build argv list for a CGI script according to spec
 * Speedup process_header_line, eliminate potential memory leak
 * Occasional spelling fixes and lint removal
 * Added REMOTE_PORT env var for CGI scripts, to allow easy ident lookups
 * Changed rfc822 time format
 * Log timeouts and broken connections
 * Fix mime suffix handling for filenames with multiple "."s
 * Initialize conn->time_last, fixes bug with rapid-fire connections
 * Performance tweak to req_write()
 * Changed http_version from float to char[8]
 * Rewrote on-the-fly directory generation; it works now
 * Added user configurable dircache directory in boa.conf
 * Fixed "simple" response bugs, including incorrect CGI handling
 * Keepalive (HTTP/1.1 draft) support, mostly by Jon Nelson
 * Close data_fd in 304 Not Modified flow of control
 * Switch socket flags to non-blocking before cgi handoff
 * Try to handle errno properly in the face of multiple errors
 * Close fd's of all other transactions before cgi handoff
 * Move real work for sighup and sigchld out of signal handler
 * Fix free(req->cgi_env) in request.c
 * Response message cleanup - better match to HTML-2.0 DTD
 * Experimental Virtual Host code from Russ Nelson <nelson@crynwr.com>
 * Expand buffer for escaped URI in init_get()
 * SIGTERM triggers lame duck mode until all pending transactions complete
 * Close and unlink temp file for POST in parent process

** Changes from v0.90 to v0.91 **
 
 * Cleaned up main while loop
 * Optimized request line parsing
 * Added state machine for header reads -- necessary to deal wtih
 possibility of obtaining header data in multiple reads.  This 
 also allows interactive use of server.
 * Added 500/501 return codes for various conditions

** v0.90 **

 * Initial release



src/defines.h
/*
 *  Boa, an http server
 *  Copyright (C) 1995 Paul Phillips <psp@well.com>
 *  Some changes Copyright (C) 1996, 1997 Larry Doolittle <ldoolitt@boa.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 1, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#ifndef _DEFINES_H
#define _DEFINES_H

/***** Change this, or use -c on the command line to specify it *****/

#define SERVER_ROOT "/usr/local/boa"

/***** Various stuff that you may want to tweak, but probably shouldn't *****/

#define SOCKETBUF_SIZE				4096
#define RESPONSEBUF_SIZE			1024
#define MAX_HEADER_LENGTH			SOCKETBUF_SIZE

#define MIME_HASHTABLE_SIZE			47
#define ALIAS_HASHTABLE_SIZE			17
#define PASSWD_HASHTABLE_SIZE			47
#define DNS_HASHTABLE_SIZE			117

#define REQUEST_TIMEOUT				60

#define CGI_MIME_TYPE    "application/x-httpd-cgi"
#define DEFAULT_PATH     "/bin:/usr/bin:/usr/ucb"

/***** CHANGE ANYTHING BELOW THIS LINE AT YOUR OWN PERIL *****/
/***** You will probably introduce buffer overruns unless you know
       what you are doing *****/

#define TIMEBUF_SIZE				64

#define MAX_SITENAME_LENGTH			256
#define MAX_CGI_VARS				50
#define MAX_LOG_LENGTH				MAX_HEADER_LENGTH + 1024
#define MAX_FILE_LENGTH				256
#define MAX_PATH_LENGTH				512

#define COMMON_CGI_VARS				8

#define SERVER_VERSION				"Boa/0.92r"
#define CGI_VERSION				"CGI/1.1"

/******************* RESPONSE CLASSES *****************/

#define R_INFORMATIONAL	1
#define R_SUCCESS	2
#define R_REDIRECTION	3	
#define R_CLIENT_ERROR	4
#define R_SERVER_ERROR	5

/******************* RESPONSE CODES ******************/

#define R_REQUEST_OK	200
#define R_CREATED	201
#define R_ACCEPTED	202
#define R_PROVISIONAL	203	/* provisional information */
#define R_NO_CONTENT	204

#define R_MULTIPLE	300	/* multiple choices */
#define R_MOVED_PERM	301
#define R_MOVED_TEMP	302
#define R_NOT_MODIFIED	304

#define R_BAD_REQUEST	400
#define R_UNAUTHORIZED	401
#define R_PAYMENT	402	/* payment required */
#define R_FORBIDDEN	403
#define R_NOT_FOUND	404
#define R_METHOD_NA	405	/* method not allowed */
#define R_NONE_ACC	406	/* none acceptable */
#define R_PROXY		407	/* proxy authentication required */
#define R_REQUEST_TO	408	/* request timeout */
#define R_CONFLICT	409
#define R_GONE		410

#define R_ERROR		500	/* internal server error */
#define	R_NOT_IMP	501	/* not implemented */
#define	R_BAD_GATEWAY	502	
#define R_SERVICE_UNAV	503	/* service unavailable */
#define	R_GATEWAY_TO	504	/* gateway timeout */

/****************** METHODS *****************/

#define M_GET		1 
#define M_HEAD		2
#define M_PUT		3
#define M_POST		4
#define M_DELETE	5
#define M_LINK		6
#define M_UNLINK	7

/******************* ERRORS *****************/

#define SERVER_ERROR		1
#define OUT_OF_MEMORY		2
#define NO_CREATE_SOCKET	3
#define NO_FCNTL		4
#define NO_SETSOCKOPT		5
#define NO_BIND			6
#define NO_LISTEN		7
#define NO_SETGID		8
#define NO_SETUID		9
#define NO_OPEN_LOG		10
#define NO_CREATE_TMP		11
#define WRONG_TMP_STAT		12

/************** REQUEST STATUS (req->status) ***************/

#define WRITE                   0
#define READ_HEADER             1
#define READ_BODY               2
#define ONE_CR			3
#define ONE_LF			4
#define TWO_CR			5

/************** CGI STATUS (req->is_cgi) ******************/

#define CGI                     1
#define NPH                     2

/************** ALIAS TYPES (aliasp->type) ***************/

#define ALIAS			0
#define SCRIPTALIAS		1
#define REDIRECT		2

#endif
